### PR TITLE
MueLu: HHG Structured Region Driver

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/muelu 
@rstumin @mayrmt @pohm01 

## Description
Adding driver for fully structured regions

## Motivation and Context
We need this driver to check that the HHG implementation is correct.

## Related Issues

* Blocks 

## How Has This Been Tested?
The code builds and does not introduces warnings.
It is not meant to be used at this team and is protected in an experimental folder

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.

### CODE REVIEW

File: packages/muelu/research/luc/region_algorithms/Driver_Structured_Regions.cpp [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+    galeriList.set("back boundary"  , "Neumann");
+  }
+
+  RCP<Galeri::Xpetra::Problem<Map,CrsMatrixWrap,MultiVector> > Pr =
+    Galeri::Xpetra::BuildProblem<SC,LO,GO,Map,CrsMatrixWrap,MultiVector>(galeriParameters.GetMatrixType(), map, galeriList);
+  A = Pr->BuildMatrix();
+
+  if(matrixType == "Elasticity2D") {
+    numDofsPerNode = 2;
+  } else if(matrixType == "Elasticity2D") {
```
[the remark this answers is not in the document]
Comment (lucbv) [reply packages/muelu/research/luc/region_algorithms/Driver_Structured_Regions.cpp]: Obviously! Just, took care of it.

**Comment by lucbv** [comment]
@mayrmt, you are right, I was not using correct logic for the `sendGIDs`.
I am fixing it and should be able to push an update today.

**Comment by mayrmt** [comment]
@lucbv I was just double-checking a larger example, i.e. 10x10 nodes on 9 processors. Here's the output:

```
mpirun -np 9 ./MueLu_StructuredRegionsDriver.exe 
========================================================
Xpetra::Parameters
 Linear algebra library: Tpetra
Galeri::Xpetra::Parameters<int>
 Matrix type: Laplace2D
 Problem size: 100 (10x10)
Processor subdomains in x direction: 3
Processor subdomains in y direction: 3
Processor subdomains in z direction: -1
========================================================
Galeri complete.
========================================================
p=0 | startGID= 0, startIndices: {0, 0, 0}, endIndices: {3, 3, 0}, gNodesPerDim: {10, 10, 1}, BCs={1, 0, 1, 0, 1, 1}
p=0 | numReceive=0, numSend=9
p=0 | receiveGIDs: {}
p=0 | receivePIDs: {}
p=0 | sendGIDs: {3, 13, 23, 33, 30, 31, 32, 33, 33}
p=0 | sendPIDs: {1, 1, 1, 1, 3, 3, 3, 3, 4}
p=1 | startGID= 4, startIndices: {4, 0, 0}, endIndices: {6, 3, 0}, gNodesPerDim: {10, 10, 1}, BCs={0, 0, 1, 0, 1, 1}
p=1 | numReceive=4, numSend=7
p=1 | receiveGIDs: {3, 13, 23, 33}
p=1 | receivePIDs: {0, 0, 0, 0}
p=1 | sendGIDs: {6, 16, 26, 36, 35, 36, 36}
p=1 | sendPIDs: {2, 2, 2, 2, 4, 4, 5}
p=2 | startGID= 7, startIndices: {7, 0, 0}, endIndices: {9, 3, 0}, gNodesPerDim: {10, 10, 1}, BCs={0, 1, 1, 0, 1, 1}
p=2 | numReceive=4, numSend=3
p=2 | receiveGIDs: {6, 16, 26, 36}
p=2 | receivePIDs: {1, 1, 1, 1}
p=2 | sendGIDs: {37, 38, 39}
p=2 | sendPIDs: {5, 5, 5}
p=3 | startGID= 40, startIndices: {0, 4, 0}, endIndices: {3, 6, 0}, gNodesPerDim: {10, 10, 1}, BCs={1, 0, 0, 0, 1, 1}
p=3 | numReceive=4, numSend=7
p=3 | receiveGIDs: {30, 31, 32, 33}
p=3 | receivePIDs: {0, 0, 0, 0}
p=3 | sendGIDs: {53, 63, 60, 61, 62, 63, 63}
p=3 | sendPIDs: {4, 4, 6, 6, 6, 6, 7}
p=6 | startGID= 70, startIndices: {0, 7, 0}, endIndices: {3, 9, 0}, gNodesPerDim: {10, 10, 1}, BCs={1, 0, 0, 1, 1, 1}
p=6 | numReceive=4, numSend=3
p=6 | receiveGIDs: {60, 61, 62, 63}
p=6 | receivePIDs: {3, 3, 3, 3}
p=6 | sendGIDs: {73, 83, 93}
p=6 | sendPIDs: {7, 7, 7}
p=7 | startGID= 74, startIndices: {4, 7, 0}, endIndices: {6, 9, 0}, gNodesPerDim: {10, 10, 1}, BCs={0, 0, 0, 1, 1, 1}
p=7 | numReceive=7, numSend=3
p=7 | receiveGIDs: {63, 64, 65, 66, 73, 83, 93}
p=7 | receivePIDs: {3, 4, 4, 4, 6, 6, 6}
p=7 | sendGIDs: {76, 86, 96}
p=7 | sendPIDs: {8, 8, 8}
p=8 | startGID= 77, startIndices: {7, 7, 0}, endIndices: {9, 9, 0}, gNodesPerDim: {10, 10, 1}, BCs={0, 1, 0, 1, 1, 1}
p=8 | numReceive=7, numSend=0
p=8 | receiveGIDs: {66, 67, 68, 69, 76, 86, 96}
p=8 | receivePIDs: {4, 5, 5, 5, 7, 7, 7}
p=8 | sendGIDs: {}
p=8 | sendPIDs: {}
p=4 | startGID= 44, startIndices: {4, 4, 0}, endIndices: {6, 6, 0}, gNodesPerDim: {10, 10, 1}, BCs={0, 0, 0, 0, 1, 1}
p=4 | numReceive=7, numSend=5
p=4 | receiveGIDs: {33, 34, 35, 36, 43, 53, 63}
p=4 | receivePIDs: {0, 1, 1, 1, 3, 3, 3}
p=4 | sendGIDs: {56, 66, 65, 66, 66}
p=4 | sendPIDs: {5, 5, 7, 7, 8}
p=5 | startGID= 47, startIndices: {7, 4, 0}, endIndices: {9, 6, 0}, gNodesPerDim: {10, 10, 1}, BCs={0, 1, 0, 0, 1, 1}
p=5 | numReceive=7, numSend=3
p=5 | receiveGIDs: {36, 37, 38, 39, 46, 56, 66}
p=5 | receivePIDs: {1, 2, 2, 2, 4, 4, 4}
p=5 | sendGIDs: {67, 68, 69}
p=5 | sendPIDs: {8, 8, 8}
===========================================================================================================
```
Most of it looks as I would expect it, though here are some irregularities:
- On proc 1, sendGID 34 seems to be missing:
    ```
    p=1 | sendGIDs: {6, 16, 26, 36, 35, 36, 36} 
    p=1 | sendPIDs: {2, 2, 2, 2, 4, 4, 5}
    ```
    It should be 
    ```
    p=1 | sendGIDs: {6, 16, 26, 36, 34, 35, 36, 36} 
    p=1 | sendPIDs: {2, 2, 2, 2, 4, 4, 4, 5}
    ```
- On proc 3, sendGID 43 seems to be missing:
    ```
    p=3 | sendGIDs: {53, 63, 60, 61, 62, 63, 63}
    p=3 | sendPIDs: {4, 4, 6, 6, 6, 6, 7}
    ````
    It should be
    ```
    p=3 | sendGIDs: {43, 53, 63, 60, 61, 62, 63, 63}
    p=3 | sendPIDs: {4, 4, 4, 6, 6, 6, 6, 7}
    ```
- On proc 4, sendGIDs 46 and 64 are missing:
    ```
    p=4 | sendGIDs: {56, 66, 65, 66, 66}
    p=4 | sendPIDs: {5, 5, 7, 7, 8}
    ```
    It should be
    ```
    p=4 | sendGIDs: {46, 56, 66, 64, 65, 66, 66}
    p=4 | sendPIDs: {5, 5, 5, 7, 7, 7, 8}
    ```

I'll post a little sketch with all the GIDs as I expect them to be in a minute.

**Comment by mayrmt** [comment]
Here it is: [MueLu_HHG.pdf](https://github.com/trilinos/Trilinos/files/2832652/MueLu_HHG.pdf)

**Comment by lucbv** [comment]
@mayrmt, thank for testing this, I'll double check what's going on!
The way this thing is implemented leaves a lot of room for improvement since I am basically just using a gigantic if statement to catch all possible cases. Something more automated would probably be more reliable...

**Pull request merged**